Modified template indexing to broadcast indices to out instead of mask and some other flexattention micro-opts

Chillee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129013
* __->__ #128938

For headdim=64 and headdim=128

Old:
<img width="656" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/2c5d1613-96dc-4300-8dc0-dccaef59e73c">

New:
<img width="644" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/730004a8-6d5f-46a5-82a0-2594feb5e192">


Note, this does regress headdim=256. We can unregress it by special casing `headdim=256`, but ehh.... we can do it later

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Chillee committed:
Modified template indexing to broadcast indices to out instead of mask and some other flexattention micro-opts

[ghstack-poisoned]

Chillee committed:
Update on "Modified template indexing to broadcast indices to out instead of mask and some other flexattention micro-opts"


For headdim=64 and headdim=128

Old:
<img width="656" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/2c5d1613-96dc-4300-8dc0-dccaef59e73c">

New:
<img width="644" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/730004a8-6d5f-46a5-82a0-2594feb5e192">


Note, this does regress headdim=256. We can unregress it by special casing `headdim=256`, but ehh....

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -265,24 +263,26 @@ def build_subgraph_buffer(
         # -- compute scaling constant ---
-        row_max = tl.max(post_mod_scores, 1)
-        m_i_new = tl.maximum(m_i, row_max)
+        m_ij = tl.maximum(m_i, tl.max(post_mod_scores, 1))
```
does this actually have an effect on perf?

IMO its more readable with row_max

Chillee replied:
No, I just wanted to mimic the triton tutorial more lol.

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -265,24 +263,26 @@ def build_subgraph_buffer(
+        # # -- scale and update acc --
+        acc = acc * alpha[:, None]
+        v = tl.load(V_block_ptr)
+        acc = tl.dot(p.to(MATMUL_PRECISION), v, acc)
```
do you need to cast v.to(MM_precision) anymore?

Chillee replied:
no, because `v` is directly loaded and should already be in the right precision.

drisspg commented on torch/_inductor/kernel/flex_attention.py:
```diff
@@ -294,8 +294,8 @@ def build_subgraph_buffer(
     idx_m = offs_m[:, None]
     idx_d = tl.arange(0, BLOCK_DMODEL)[None, :]
 
+    mask = idx_m < Q_LEN
```
love it

drisspg commented on torch/_inductor/select_algorithm.py:
```diff
@@ -147,21 +157,34 @@ def __init__(
         self.subgraphs: Optional[List[ir.ComputedBuffer]] = subgraphs
+
+        # The following attributes (body, template_mask, output_val) are all
+        # used for triton kernel codegen.
```
nit: maybe write a note around the named_tuple and point back there, describing what these things are used for

Chillee replied:
Doing in a follow-up PR

drisspg reviewed (approved):
Some nits

Chillee committed:
Update on "Modified template indexing to broadcast indices to out instead of mask and some other flexattention micro-opts"


For headdim=64 and headdim=128

Old:
<img width="656" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/2c5d1613-96dc-4300-8dc0-dccaef59e73c">

New:
<img width="644" alt="image" src="https://github.com/pytorch/pytorch/assets/6355099/730004a8-6d5f-46a5-82a0-2594feb5e192">


Note, this does regress headdim=256. We can unregress it by special casing `headdim=256`, but ehh.... we can do it later

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Chillee commented:
@pytorchbot merge -i

Chillee commented:
@pytorchbot merge -i
